server/controllers/joinLeaveCancelController.js

```diff
@@ -0,0 +1,144 @@
+              status: 409,
+              message: "You are already registered in this event.",
+            });
+          }
+          //if you don't find a matching participant.
+          //add the incoming participant to that.
+          let updateParticipant = await addParticipant(
+            eventInformation.participantId,
+            participantDetails
+          );
+          //when signing up, you create a _id = to the userId in the collectionUserEvents.
+          //now we will push the event id in the array in this event to keep track of which events the user JOINED!
+          let updateUserEvent = await addUserEvent(
+            participantDetails.userId,
+            participantDetails.eventId
+          );
+          if (updateUserEvent && updateParticipant) {
+            res
+              .status(204)
+              .json({ status: 204, message: "Successfully joined the event!" });
+          }
+        }
+      }
+    } else {
+      res.status(400).json({
+        status: 400,
+        message:
+          "Information was not received correctly. Try refreshing the page.",
+      });
+    }
+  } catch (error) {
+    console.log(error.stack, "Catch Error in handleJoinEvent");
+    res.status(500).json({ status: 500, message: error.message });
+  }
+};
+//@endpoint POST /leaveEvent
+//@desc leave the event selected from ViewActivity comp.
+//@access PRIVATE - will need to validate token? YES
+const handleLeaveEvent = async (req, res, next) => {
+  const participantDetails = req.body.participantDetails;
+  const eventInformation = req.body.eventInformation;
+  try {
+    //now you have the participant ID from eventInformation.
+    //find the participant collect from the event ParticipantId key.
+    //now within that, remove the participant
+    let updateParticipant = await removeParticipant(
+      eventInformation.participantId,
+      participantDetails.userId
+    );
+    let updateUserEvent = await removeUserEvent(
+      participantDetails.userId,
+      eventInformation._id
+    );
+    if (updateParticipant && updateUserEvent) {
+      res
+        .status(200)
+        .json({ status: 200, message: "Successfully left the event!" });
+    }
+  } catch (error) {
+    console.log(error.stack, "Catch Error in handleLeaveEvent");
+    res.status(500).json({ status: 500, message: error.message });
+  }
+};
+
+//@endpoint POST /cancelEvent
+//@desc cancel the event selected
+//@access PRIVATE - will need to validate token? YES
+const handleCancelEvent = async (req, res, next) => {
+  const eventId = req.body.eventId;
+  try {
+    let getEvent = await getEventById(eventId); //get the event for participantId.
+    let deletedRoom = await deleteRoom(eventId); //for chat system.
+    let deletedEvent = await deleteEvent(eventId); //delet
+    let getParticipants = await getParticipantsById(getEvent.participantId);
+    let removeUserAsParticipant = await getParticipants.participants.forEach(
```

Comment:
This can be refactored to just one mongo query, rather than using a loop here. I wouldn't address now, but you can add a todo later to optimize